### ISSUE TITLE
[Kernel] SmolVLA expert kernels; GPT-2 remains a fixture

### ISSUE THREAD

**Issue by elprofesoriqo** [elprofesoriqo/FlowEdge#10]
## Goal

Add a causal Transformer backbone behind the same interface as the existing Mamba backbone.

The first implementation should support a fixed-shape GPT-2 style Transformer with preallocated KV cache and streaming decode.

Reference model:

* `openai-community/gpt2`

## Architecture

```mermaid
flowchart TD
    A[Input prefix] --> B[Token embedding]
    B --> C[Position embedding]
    C --> D[Transformer Block 1]
    D --> E[Transformer Block 2]
    E --> F[...]
    F --> G[Transformer Block N]
    G --> H[Final LayerNorm]
    H --> I[Conditioning vector]
    I --> J[FlowEdge action head]
```

### Transformer block

```mermaid
flowchart TD
    A[Input] --> B[LayerNorm]
    B --> C[Q K V projections]
    C --> D[Causal multi-head attention]
    D --> E[Output projection]

    A --> F[Residual add]
    E --> F

    F --> G[LayerNorm]
    G --> H[MLP projection]
    H --> I[GELU]
    I --> J[MLP output projection]

    F --> K[Residual add]
    J --> K

    K --> L[Block output]
```

## Initial scope

Support:

* causal multi-head self-attention
* LayerNorm
* GELU MLP
* fixed maximum prefix length
* batch size `1`
* preallocated per-layer KV cache
* full-prefix inference
* streaming decode
* GPT-2 style learned position embeddings
* FP32
* BF16 where existing kernel support allows it
* output through the existing backbone conditioning interface

## Out of scope

Not required for the first implementation:

* generic Hugging Face Transformer support
* Llama, Gemma, Qwen, Mistral
* RoPE
* grouped-query attention
* mixture-of-experts
* training
* text generation API
* dynamic KV-cache growth
* arbitrary batch sizes
* accelerator-specific attention kernels

Additional Transformer families should be added separately.

## Kernel primitives

Add the required kernels behind `kernels.h`:

* [ ] stable softmax
* [ ] LayerNorm
* [ ] GELU
* [ ] scaled dot-product attention
* [ ] causal masking

Start with correct scalar implementations.

AVX2 and NEON implementations can follow once the scalar path is tested and benchmarked.

Requirements:

* no internal heap allocation
* deterministic results for fixed inputs
* reference tests for every primitive

## Multi-head attention

Implement causal multi-head attention:

* [ ] Q projection
* [ ] K projection
* [ ] V projection
* [ ] head reshape
* [ ] scaled dot-product attention
* [ ] causal masking
* [ ] softmax
* [ ] value aggregation
* [ ] output projection

Initial constraints:

```text
batch_size = 1
sequence_length <= max_prefix
hidden_dim % num_heads == 0
```

## KV cache

Allocate all KV-cache memory during model initialization.

```mermaid
flowchart LR
    A[New token] --> B[Q K V projection]

    B --> C[Query]
    B --> D[New key]
    B --> E[New value]

    D --> F[(K cache)]
    E --> G[(V cache)]

    C --> H[Causal attention]
    F --> H
    G --> H

    H --> I[Block output]
```

Requirements:

* [ ] per-layer K cache
* [ ] per-layer V cache
* [ ] fixed capacity based on `max_prefix`
* [ ] cache allocated from the FlowEdge arena
* [ ] current decode position tracked explicitly
* [ ] cache reset support
* [ ] no allocation during decode
* [ ] bounds checking when `max_prefix` is reached

## Transformer backbone

Add:

```text
models/transformer/
├── transformer.h
└── transformer.cc
```

The backbone should support:

* [ ] model configuration loaded at initialization
* [ ] all working buffers allocated at initialization
* [ ] full-prefix forward
* [ ] streaming decode
* [ ] final conditioning vector compatible with FlowEdge action heads

## Streaming decode

Streaming decode should reuse the KV cache instead of recomputing previous keys and values.

```mermaid
flowchart TD
    A[Token t] --> B[Transformer decode]
    B --> C[Append K and V at position t]
    C --> D[Attend to positions 0..t]
    D --> E[Hidden state t]
    E --> F[Conditioning state]
```

Requirements:

* [ ] one new token processed per decode call
* [ ] previous K/V state reused
* [ ] previous K/V projections are not recomputed
* [ ] no heap allocation after initialization

## Backbone selection

Extend model loading to support:

```text
mamba
transformer
```

Prefer explicit checkpoint metadata over tensor-name inference.

Requirements:

* [ ] existing Mamba checkpoints continue to work
* [ ] Transformer checkpoints load correctly
* [ ] unsupported backbone values fail with a clear error

## Model conversion

Extend `convert/` with one explicit mapping:

```mermaid
flowchart LR
    A[Hugging Face GPT-2] --> B[FlowEdge converter]

    B --> C[Embeddings]
    B --> D[Attention weights]
    B --> E[LayerNorm]
    B --> F[MLP weights]
    B --> G[Model metadata]

    C --> H[FlowEdge safetensors]
    D --> H
    E --> H
    F --> H
    G --> H

    H --> I[Transformer loader]
```

The converter should:

* [ ] map token embeddings
* [ ] map position embeddings
* [ ] map Q/K/V projection weights
* [ ] map attention output projection
* [ ] map LayerNorm parameters
* [ ] map MLP projections
* [ ] map final LayerNorm
* [ ] store Transformer configuration metadata
* [ ] validate hidden size, head count, layer count, and prefix limit
* [ ] reject unsupported variants

Do not treat this as generic HF causal-LM conversion.

## Verification

Extend `torch_ref.py` with the Transformer reference path.

Verify:

* [ ] GELU
* [ ] LayerNorm
* [ ] softmax
* [ ] attention output
* [ ] Transformer block output
* [ ] full-prefix output
* [ ] streaming output
* [ ] final conditioning vector

### Full-prefix vs streaming parity

```mermaid
flowchart TD
    A[Same input sequence]

    A --> B[Full-prefix forward]
    A --> C[Streaming decode]

    C --> D[decode token 0]
    D --> E[decode token 1]
    E --> F[...]
    F --> G[decode token N]

    B --> H[Full-prefix output]
    G --> I[Streaming output]

    H --> J{Within tolerance}
    I --> J

    J -->|Yes| K[Pass]
    J -->|No| L[Fail]
```

## Tests

* [ ] GELU vs reference
* [ ] LayerNorm vs reference
* [ ] softmax vs reference
* [ ] causal mask correctness
* [ ] single-head attention vs naive reference
* [ ] multi-head attention vs PyTorch
* [ ] KV-cache update
* [ ] KV-cache reset
* [ ] `max_prefix` bounds handling
* [ ] full-prefix Transformer parity
* [ ] streaming Transformer parity
* [ ] full-prefix vs streaming parity
* [ ] GPT-2 conversion
* [ ] backbone selection
* [ ] no-allocation streaming decode

## Performance requirements

After initialization:

* no heap allocation during forward or decode
* KV cache stays fixed-size
* intermediate buffers are reused
* streaming decode reuses previous K/V state

Benchmark:

* full-prefix forward
* single-token decode
* attention kernel
* complete conditioning latency

Report:

* mean
* p50
* p95
* p99

Include:

* CPU
* OS
* compiler
* build type
* model dimensions
* sequence length
* precision

## Acceptance criteria

* [ ] GPT-2 reference checkpoint converts successfully
* [ ] converted model loads in FlowEdge
* [ ] full-prefix output matches the PyTorch reference within documented tolerance
* [ ] streaming output matches the PyTorch reference within documented tolerance
* [ ] full-prefix and streaming results agree
* [ ] KV caches are allocated during initialization
* [ ] decode performs no heap allocation
* [ ] `max_prefix` is enforced
* [ ] existing Mamba models continue to work
* [ ] Transformer output works with the existing backbone interface
* [ ] tests pass
* [ ] supported scope and limitations are documented

## Suggested child issues

* [ ] `[Kernel] Add GELU, LayerNorm and stable softmax`
* [ ] `[Kernel] Add causal multi-head attention`
* [ ] `[Runtime] Add fixed-size Transformer KV cache`
* [ ] `[Backbone] Implement GPT-2 style Transformer backbone`
* [ ] `[Convert] Add GPT-2 converter and PyTorch parity`

## Dependency

This work provides the Transformer primitives required by #11.

## References

* `openai-community/gpt2`
* #11 π0 action expert

**Comment by elprofesoriqo** [comment]
GPT-2 product scope is closed. The remaining work is RoPE/GQA/SwiGLU helpers for the SmolVLA action expert; `tiny-gpt2` stays a conversion and streaming fixture. Child: the kernel-lift issue opened with this batch.

**Comment by elprofesoriqo** [comment]
Kernel lift tracked in #147. GPT-2 stays a fixture.

**Comment by elprofesoriqo** [comment]
Status on `feat/issue-10-transformer-cuda` (do not close until HF re-run of the fixture is on a machine that has the checkpoint):

Done:

- `cached_causal_attention` in `kernels.h`; Transformer decode uses it
- gtests: GELU, LayerNorm, softmax, causal mask, KV reset, max-prefix reject, prefix vs stream, embedding mask
- `transformer_latency` for host decoder p50/p99 JSON (not a policy number)
- local gate: `verify_all.sh` / `verify_local.ps1` run HF parity when `models/tiny-gpt2` is present
- docs: tiny-gpt2 is the CI fixture; `openai-community/gpt2` is optional same `model_type: gpt2` converter
- existing artifact `bench/artifacts/transformer/tiny-gpt2-transformer-reference.json` (max abs 1.19e-7) still stands; this host did not re-download the weights

Not done, so this issue stays open:

- robotics policy adapter / action-head conditioning product path
- optional `openai-community/gpt2` conversion JSON (weights not on this workstation)
- latency JSON for tiny-gpt2 (same missing checkpoint)

SmolVLA expert kernels remain a separate remaining slice of this title.
